RandomMCPlayer.java

```diff
@@ -168,7 +170,192 @@ private int simPlay(int depthLimit) {
+		}
+		return hand;
+	}
+
+	private Card[] getHandByColumn(int col) {
+		Card[] hand = new Card[SIZE];
+		for (int row = 0; row < SIZE; row++) {
+			hand[row] = grid[row][col];
+		}
+		return hand;
+	}
+
+	private float getHandUtility(Card[] hand, boolean isRow) {
+		float utility = 0;
+		int numCards = 0;
+		for (int i = 0; i < SIZE ; i++) {
+			if (hand[i] != null) numCards++;
+		}
+		if (numCards == 0) utility = 0;
+		else if (numCards == 5) utility = system.getHandScore(hand);
+		else if (numCards == 4) utility = getExpectedValue(hand);
+		else utility = 3; // temp
+		return utility;
+	}
+
+	public float getExpectedValue(Card[] hand) {
+		float expectedValue = 0;
+		Integer [] suitArr  = new Integer[4];
+		Integer [] rankArr  = new Integer[13];
+
+		for (int i = 0; i < 5; i++) {
+			if (hand[i] != null) {
+				suitArr[hand[i].getSuit()]++;
+				rankArr[hand[i].getRank()]++;
+			}
+		}
+		ArrayList<Integer> suitArrayList = new ArrayList<>(Arrays.asList(suitArr));
+		ArrayList<Integer> rankArrayList = new ArrayList<>(Arrays.asList(rankArr));
+
+		if (suitArrayList.contains(4)){
+			expectedValue += probOfRoyalFlush(hand, suitArrayList.indexOf(4)) * 30; // royal flush
+			expectedValue += probOfSequence(hand, suitArrayList.indexOf(4)) * 30; // straight flush
+			expectedValue += probOfSuit(suitArrayList.indexOf(4)) * 5; // flush
+		}
+		
+		if (rankArrayList.contains(3)) {
+			expectedValue += probOfRank(rankArrayList.indexOf(3)) * 16; // fourOfKind
+			expectedValue += probOfRank(rankArrayList.indexOf(1)) * 10; // find card to complete 2 card rank for full-house 
+		}
+		else if (rankArrayList.contains(2)) {
+			if (Collections.frequency(rankArrayList, 2) == 2) { // check if there's two occurence of 2 
+				expectedValue += probOfRank(rankArrayList.indexOf(2), rankArrayList.lastIndexOf(2)) * 10; // find card for 3 card rank of full-house
+			}
+			else {
+				expectedValue += probOfRank(rankArrayList.lastIndexOf(2)) * 6; // threeOfKind
+				expectedValue += probOfRank(rankArrayList.lastIndexOf(1)) * 3; // twoPair
+			}
+		}
+		else {
+			expectedValue += probOfRank(rankArrayList) * 1; // onePair
+		}
+
+		expectedValue += probOfSequence(hand, -1) * 12; // straight flush
+
+		return expectedValue;
+	}
+
+	float probOfRoyalFlush(Card[] hand, int suit) {
+		int rankToGet = 0;
+		boolean correctSequence = true;
+		int undealtRoyalCardCount = 0;
+
+		for (int i=0; i < 5; i++) {
+			if (hand[i] == null ) {
+				rankToGet = (i==4)? 0: (i+10);
+			}
+			else if (hand[i].getRank() != ((i==4)? 0: (i+10))){
+				correctSequence = false;
+			}
+		}
+
+		if (correctSequence) { // probability of missing rank with specified suit
+			for (int i=numPlays; i<simDeck.length; i++) {
+				Card card = simDeck[i];
+				if (card.getSuit() == suit && card.getRank() == rankToGet) undealtRoyalCardCount++;
+			}
+			return undealtRoyalCardCount;
+		}
+		else {
+			return 0;
+		}
+	}
 	
+	float probOfSequence(Card[] hand, int suit) {
+		int rankToGet = 0;
```

Comment:
The straight (i.e., sequence) check is the most complicated. Here's the basic idea:
1. Make sure there are no pairs (i.e., all ranks are different). If not, sequence is impossible.
2. Get the min and max ranks from the hand.
3. If the difference between min and max is > 4, the sequence is impossible.
4. Otherwise, difference is either 3 or 4.
5.  if difference = 4, then the missing card is between min and max. Go through the ranks in order between min and max and figure out which one is missing.
5. If difference = 3, then the missing card can be either min-1 (except if min is ace) or max+1 (except if max is king).